daemon/qrexec-daemon.c

```diff
@@ -1323,26 +1323,49 @@ void handle_message_from_agent(void)
                 free(untrusted_service_name);
                 handle_vchan_error("recv params3(service_name)");
             }
+            service_name_len -= 1;
 
             /* sanitize start */
             ENSURE_NULL_TERMINATED(untrusted_params3.target_domain);
-            untrusted_service_name[service_name_len-1] = 0;
             sanitize_name(untrusted_params3.target_domain, "@:");
-            sanitize_name(untrusted_service_name, "+");
             if (!validate_request_id(&untrusted_params3.request_id, "MSG_TRIGGER_SERVICE3")) {
-                free(untrusted_service_name);
```

Comment:
Removing this will leak `untrusted_service_name` - `free` part handles only `service_name` which is `NULL` at this point. Similar for other cases you add in this patch below.